backend/src/main/scala/bloop/Compiler.scala

```diff
@@ -357,13 +366,78 @@ object Compiler {
 
     val uniqueInputs = compileInputs.uniqueInputs
     reporter.reportStartCompilation(previousProblems)
+    val fileManager = newFileManager
+
+    // Manually skip redundant best-effort compilations. This is necessary because compiler
+    // phases supplying the data needed to skip compilations in zinc remain unimplemented for now.
+    val noopBestEffortResult = compileInputs.previousCompilerResult match {
+      case Failed(
+            problems,
+            t,
+            elapsed,
+            _,
+            bestEffortProducts @ Some(BestEffortProducts(compileProducts, previousHash))
+          ) if isBestEffortMode =>
+        val newHash = BestEffortUtils.hashResult(
+          compileProducts.newClassesDir,
+          compileInputs.sources,
+          compileInputs.classpath
+        )
+
+        if (newHash == previousHash) {
+          reporter.processEndCompilation(
+            problems,
+            bsp.StatusCode.Error,
+            None,
+            None
+          )
+          reporter.reportEndCompilation()
+          val backgroundTasks = new CompileBackgroundTasks {
+            def trigger(
+                clientClassesObserver: ClientClassesObserver,
+                clientReporter: Reporter,
+                clientTracer: BraveTracer,
+                clientLogger: Logger
+            ): Task[Unit] = Task.defer {
+              val clientClassesDir = clientClassesObserver.classesDir
+              clientLogger.debug(s"Triggering background tasks for $clientClassesDir")
+              val firstTask = Task { BloopPaths.delete(AbsolutePath(newClassesDir)) }
```

Comment:
Could you explain a bit more in a comment why is this remove and later we copy from `compileProducts.newClassesDir`?

Reply:
Sure! Just added some comments there